core/src/View/Kokkos_ViewCommonType.hpp

```diff
@@ -0,0 +1,128 @@
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Part of Kokkos, under the Apache License v2.0 with LLVM Exceptions.
+// See https://kokkos.org/LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//@HEADER
+
+#ifndef KOKKOS_VIEWCOMMONTYPE_HPP
+#define KOKKOS_VIEWCOMMONTYPE_HPP
+
+#include <type_traits>
+#include <Kokkos_Macros.hpp>
+
+namespace Kokkos {
+namespace Impl {
+
+template <class Specialize, typename A, typename B>
+struct CommonViewValueType;
+
+template <typename A, typename B>
+struct CommonViewValueType<void, A, B> {
+  using value_type = std::common_type_t<A, B>;
+};
+
+template <class Specialize, class ValueType>
+struct CommonViewAllocProp;
+
+template <class ValueType>
+struct CommonViewAllocProp<void, ValueType> {
+  using value_type        = ValueType;
+  using scalar_array_type = ValueType;
+
+  template <class... Views>
+  KOKKOS_INLINE_FUNCTION CommonViewAllocProp(const Views&...) {}
+};
+
+template <class... Views>
+struct DeduceCommonViewAllocProp;
+
+// Base case must provide types for:
+// 1. specialize  2. value_type  3. is_view  4. prop_type
+template <class FirstView>
+struct DeduceCommonViewAllocProp<FirstView> {
+  using specialize = typename FirstView::traits::specialize;
+
+  using value_type = typename FirstView::traits::value_type;
+
+  enum : bool { is_view = is_view<FirstView>::value };
+
+  using prop_type = CommonViewAllocProp<specialize, value_type>;
+};
+
+template <class FirstView, class... NextViews>
+struct DeduceCommonViewAllocProp<FirstView, NextViews...> {
+  using NextTraits = DeduceCommonViewAllocProp<NextViews...>;
+
+  using first_specialize = typename FirstView::traits::specialize;
+  using first_value_type = typename FirstView::traits::value_type;
+
+  enum : bool { first_is_view = is_view<FirstView>::value };
+
+  using next_specialize = typename NextTraits::specialize;
+  using next_value_type = typename NextTraits::value_type;
+
+  enum : bool { next_is_view = NextTraits::is_view };
+
+  // common types
+
+  // determine specialize type
+  // if first and next specialize differ, but are not the same specialize, error
+  // out
+  static_assert(!(!std::is_same_v<first_specialize, next_specialize> &&
+                  !std::is_void_v<first_specialize> &&
+                  !std::is_void_v<next_specialize>),
+                "Kokkos DeduceCommonViewAllocProp ERROR: Only one non-void "
+                "specialize trait allowed");
+
+  // otherwise choose non-void specialize if either/both are non-void
+  using specialize =
+      std::conditional_t<std::is_same_v<first_specialize, next_specialize>,
+                         first_specialize,
+                         std::conditional_t<(std::is_void_v<first_specialize> &&
+                                             !std::is_void_v<next_specialize>),
```

Comment:
You don't need to check both `std::is_void_v<first_specialize>` and `!std::is_void_v<next_specialize>`. We already know that they have to be the same because of the `static_assert`.

Reply:
I can adjust this but this is just moving the existing implementation so I didn't want to change the implementation in a drive-by commit since it was unrelated